### ISSUE TITLE
PHPUnit error, please help!

### ISSUE THREAD

**Issue by mpresland** [tomwalder/php-gds#103]
Hi Tom,

I have just tried a unit test on a new project using PHP-GDS, when running tests I get the following message:
Exception: Could not determine DATASET, please pass to GDS\Gateway\ProtoBuf::__construct()
Are you able to help me with this? My code works in the browser, but I don't seem to be able to run tests on it.

**Comment by tomwalder** [comment]
When running the software in an AppEngine environment (local development or on production) it auto-detects the DATASET from the `$_SERVER` environment variable.

If you are running unit tests, you will need to configure it directly (on construction of the Gateway) or define it in the `$_SERVER` array.

See here:
https://github.com/tomwalder/php-gds/blob/master/src/GDS/Gateway/ProtoBuf.php#L70

**Comment by mpresland** [comment]
Hi Tom,

Thanks for the quick response! I have configured the DATASET using the $_SERVER array, and am now able to create gateways and instances of stores, however anytime I try to read/write to/from the datastore in unit tests I get the following failure:

google\appengine\runtime\RPCFailedError: Remote implementation for datastore_v4.Lookup failed

Fatal error: Uncaught exception 'google\appengine\runtime\RPCFailedError' with message 'Remote implementation for datastore_v4.Commit failed' in /Users/mattpresland/Dropbox/Projects/Web/buzilog/vendor/google/appengine-php-sdk/google/appengine/runtime/VmApiProxy.php:179

Do you have any idea what might be causing this? It is driving me nuts! I have used your php-gds library on a couple of small tasks, but I am trying to get a new project underway with full PHPUnit tests and it seems the local GAE SDK is not very PHPUnit friendly! 

Any help would be appreciated.

Cheers,

Matt

**Comment by tomwalder** [comment]
So, the php-gds library itself has (mostly) full unit tests. 93% coverage (not that that means the tests are good... but they are OK!).

And you are right, setting up unit tests for GAE is not straight forward.

But feel free to look at how I have done it. You will need to pull in a Google repo or two.

https://github.com/tomwalder/php-gds/blob/master/composer.json#L20

And also check out my tests for some guidance!

https://github.com/tomwalder/php-gds/blob/master/tests/ProtoBufFetchTest.php#L39

You might need to set up a fake protocol buffer gateway....

**Comment by mpresland** [comment]
Thanks for your help Tom!

I will have a look at setting this up! I appreciate your help, and the PHP-GDS package!

Cheers,

Matt

**Comment by tomwalder** [comment]
No problem. I think this issue could be migrated to a feature request...

"Test harness for downstream projects"

By which I mean a set of fake Gateway classes that allow unit tests to run. Perhaps configured to respond with pre-defined data sets during test runs.

**Comment by mpresland** [comment]
That would be great! Currently I have set up a very basic mock gateway class that just intercepts the upsert/delete/fetch calls and uses an internal array of entities to store and return data. This doesn't let me run any queries though! But it did help me test some basics and ensure data was being passed to the gateway correctly and processed correctly by my app on its return.

I would love to see a more robust testing gateway though! Thanks again for your help

**Comment by mpresland** [comment]
Hi Tom, 

I have been "away" from developing on GAE for a while, and just tried to fire up an application on the local dev server. It seems there have been many updates since I last used the dev server! I am currently not able to run my app, I get an error that says:

'Illegal string "dev~None" in dataset id.'

I know when I updated the SDK it made me remove my application ID etc from the .yaml, do I now need to specify the AppID or Dataset ID somewhere to access the dev datastore now? I have tried to specify using the $_SERVER array as you previously advised when trying to unit test, however still no joy! 

It doesn't pay to take 5 minutes away!
